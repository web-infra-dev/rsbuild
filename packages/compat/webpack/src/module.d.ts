declare module 'webpack/lib/MultiStats.js';
