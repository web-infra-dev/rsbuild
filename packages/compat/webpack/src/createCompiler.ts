import { type Rspack, logger } from '@rsbuild/core';
import {
  type RspackConfig,
  type Stats,
  debug,
  isDev,
  onCompileDone,
} from '@rsbuild/shared';
// @ts-expect-error
import WebpackMultiStats from 'webpack/lib/MultiStats.js';
import { type InitConfigsOptions, initConfigs } from './initConfigs';
import {
  type InternalContext,
  formatStats,
  getDevMiddleware,
  getStatsOptions,
} from './shared';
import type { WebpackConfig } from './types';

export async function createCompiler({
  context,
  webpackConfigs,
}: {
  context: InternalContext;
  webpackConfigs: WebpackConfig[];
}) {
  debug('create compiler');
  await context.hooks.onBeforeCreateCompiler.call({
    bundlerConfigs: webpackConfigs as RspackConfig[],
  });

  const { default: webpack } = await import('webpack');

  const compiler = (webpackConfigs.length === 1
    ? webpack(webpackConfigs[0])
    : webpack(webpackConfigs)) as unknown as
    | Rspack.Compiler
    | Rspack.MultiCompiler;

  const done = async (stats: unknown) => {
    const { message, level } = formatStats(
      stats as Stats,
      getStatsOptions(compiler),
    );

    if (level === 'error') {
      logger.error(message);
    }
    if (level === 'warning') {
      logger.warn(message);
    }

    if (isDev()) {
      await context.hooks.onDevCompileDone.call({
        isFirstCompile,
        stats: stats as Stats,
      });
    }

    isFirstCompile = false;
  };

  let isFirstCompile = true;

  onCompileDone(compiler, done, WebpackMultiStats);

  await context.hooks.onAfterCreateCompiler.call({
    compiler,
  });
  debug('create compiler done');

  return compiler;
}

export async function createDevMiddleware(
  options: InitConfigsOptions,
  customCompiler?: Rspack.Compiler | Rspack.MultiCompiler,
) {
  let compiler: Rspack.Compiler | Rspack.MultiCompiler;
  if (customCompiler) {
    compiler = customCompiler;
  } else {
    const { webpackConfigs } = await initConfigs(options);
    compiler = await createCompiler({
      context: options.context,
      webpackConfigs,
    });
  }

  return {
    devMiddleware: await getDevMiddleware(compiler),
    compiler,
  };
}
