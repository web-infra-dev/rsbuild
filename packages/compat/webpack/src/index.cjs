/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
