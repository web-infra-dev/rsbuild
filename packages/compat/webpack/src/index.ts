export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
