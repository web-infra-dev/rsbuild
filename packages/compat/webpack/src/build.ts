import { logger } from '@rsbuild/core';
import type { BuildOptions, Rspack } from '@rsbuild/core';
import type { Configuration as WebpackConfig } from 'webpack';
import WebpackMultiStats from 'webpack/lib/MultiStats.js';
import { createCompiler } from './createCompiler';
import { type InitConfigsOptions, initConfigs } from './initConfigs';
import { onCompileDone } from './shared';

export const build = async (
  initOptions: InitConfigsOptions,
  { mode = 'production', watch, compiler: customCompiler }: BuildOptions = {},
) => {
  if (!process.env.NODE_ENV) {
    process.env.NODE_ENV = mode;
  }

  const { context } = initOptions;

  let compiler: Rspack.Compiler | Rspack.MultiCompiler;
  let bundlerConfigs: WebpackConfig[] | undefined;

  if (customCompiler) {
    compiler = customCompiler;
  } else {
    const { webpackConfigs } = await initConfigs(initOptions);
    compiler = await createCompiler({
      context,
      webpackConfigs,
    });

    // assign webpackConfigs
    bundlerConfigs = webpackConfigs;
  }

  let isFirstCompile = true;
  await context.hooks.onBeforeBuild.call({
    bundlerConfigs: bundlerConfigs as Rspack.Configuration[],
    environments: context.environments,
    isWatch: Boolean(watch),
  });

  const onDone = async (stats: Rspack.Stats | Rspack.MultiStats) => {
    const p = context.hooks.onAfterBuild.call({
      isFirstCompile,
      stats,
      environments: context.environments,
      isWatch: Boolean(watch),
    });
    isFirstCompile = false;
    await p;
  };

  onCompileDone(compiler, onDone, WebpackMultiStats);

  if (watch) {
    compiler.watch({}, (err) => {
      if (err) {
        logger.error(err);
      }
    });
    return;
  }

  await new Promise<{ stats: Rspack.Stats | Rspack.MultiStats }>(
    (resolve, reject) => {
      compiler.run((err, stats) => {
        if (err || stats?.hasErrors()) {
          const buildError = err || new Error('Webpack build failed!');
          reject(buildError);
        }
        // If there is a compilation error, the close method should not be called.
        // Otherwise bundler may generate an invalid cache.
        else {
          // When using run or watch, call close and wait for it to finish before calling run or watch again.
          // Concurrent compilations will corrupt the output files.
          compiler.close((closeErr) => {
            closeErr && logger.error(closeErr);

            // Assert type of stats must align to compiler.
            resolve({ stats: stats as any });
          });
        }
      });
    },
  );
};
