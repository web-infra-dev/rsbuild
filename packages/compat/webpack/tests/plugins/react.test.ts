import { expect, describe, it } from 'vitest';
import { pluginReact } from '../../../../plugin-react/src';
import { webpackProvider } from '../../src';
import { pluginBabel } from '../../src/plugins/babel';
import { createStubRsbuild } from '@rsbuild/test-helper';

describe('plugins/react', () => {
  it('should work with babel-loader', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel(), pluginReact()],
      provider: webpackProvider,
      rsbuildConfig: {},
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should work with ts-loader', async () => {
    const rsbuild = await createStubRsbuild({
      provider: webpackProvider,
      rsbuildConfig: {
        tools: {
          tsLoader: {},
        },
      },
    });
    rsbuild.addPlugins([pluginReact()]);
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should not apply react refresh when dev.hmr is false', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginReact()],
      provider: webpackProvider,
      rsbuildConfig: {
        dev: {
          hmr: false,
        },
      },
    });

    expect(await rsbuild.matchBundlerPlugin('ReactRefreshPlugin')).toBeFalsy();
  });

  it('should not apply react refresh when target is node', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginReact()],
      provider: webpackProvider,
      target: 'node',
      rsbuildConfig: {},
    });

    expect(await rsbuild.matchBundlerPlugin('ReactRefreshPlugin')).toBeFalsy();
  });

  it('should not apply react refresh when target is web-worker', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginReact()],
      provider: webpackProvider,
      target: 'web-worker',
      rsbuildConfig: {},
    });

    expect(await rsbuild.matchBundlerPlugin('ReactRefreshPlugin')).toBeFalsy();
  });
});
