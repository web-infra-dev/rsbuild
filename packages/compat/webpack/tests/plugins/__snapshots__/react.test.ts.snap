// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/react > should work with babel-loader 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/babel-loader/lib/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "usage",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/babel-loader/lib/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "usage",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugins/react > should work with ts-loader 1`] = `
{
  "cache": {
    "buildDependencies": {
      "packageJson": [
        "<ROOT>/package.json",
      ],
    },
    "cacheDirectory": "<ROOT>/node_modules/.cache/webpack",
    "name": "web-test",
    "type": "filesystem",
  },
  "context": "<ROOT>",
  "devtool": "cheap-module-source-map",
  "experiments": {
    "asyncWebAssembly": true,
  },
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
    "rules": [
      {
        "resolve": {
          "fullySpecified": false,
        },
        "test": /\\\\\\.m\\?js/,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/svg/[name].[contenthash:8].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/svg/[name].[contenthash:8].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/font/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/font/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/image/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/image/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(png\\|jpg\\|jpeg\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tiff\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/media/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/media/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(mp4\\|webm\\|ogg\\|mp3\\|wav\\|flac\\|aac\\|mov\\)\\$/i,
      },
      {
        "dependency": "url",
        "generator": {
          "filename": "static/wasm/[hash].module.wasm",
        },
        "test": /\\\\\\.wasm\\$/,
        "type": "asset/resource",
      },
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/babel-loader/lib/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "usage",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-react/lib/index.js",
                  {
                    "development": true,
                    "runtime": "classic",
                    "useBuiltIns": true,
                    "useSpread": false,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "resolve": {
          "fullySpecified": false,
        },
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/babel-loader/lib/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "usage",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-react/lib/index.js",
                  {
                    "development": true,
                    "runtime": "classic",
                    "useBuiltIns": true,
                    "useSpread": false,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "resolve": {
          "preferRelative": true,
        },
        "sideEffects": true,
        "test": /\\\\\\.css\\$/,
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/css-loader",
            "options": {
              "importLoaders": 1,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]-[hash:base64:6]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  {
                    "browsers": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "chrome >= 87",
                        "edge >= 88",
                        "firefox >= 78",
                        "safari >= 14",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                ],
              },
              "sourceMap": true,
            },
          },
        ],
      },
      {
        "resolve": {
          "preferRelative": true,
        },
        "sideEffects": true,
        "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/css-loader",
            "options": {
              "importLoaders": 3,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]-[hash:base64:6]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  {
                    "browsers": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "chrome >= 87",
                        "edge >= 88",
                        "firefox >= 78",
                        "safari >= 14",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                ],
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/resolve-url-loader",
            "options": {
              "join": [Function],
              "sourceMap": false,
            },
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/sass-loader",
            "options": {
              "implementation": "<ROOT>/packages/shared/compiled/sass",
              "sourceMap": true,
            },
          },
        ],
      },
      {
        "resolve": {
          "preferRelative": true,
        },
        "sideEffects": true,
        "test": /\\\\\\.less\\$/,
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/mini-css-extract-plugin/dist/loader.js",
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/css-loader",
            "options": {
              "importLoaders": 2,
              "modules": {
                "auto": true,
                "exportLocalsConvention": "camelCase",
                "localIdentName": "[path][name]__[local]-[hash:base64:6]",
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/postcss-loader",
            "options": {
              "postcssOptions": {
                "plugins": [
                  [Function],
                  {
                    "browsers": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "info": [Function],
                    "options": {
                      "flexbox": "no-2009",
                      "overrideBrowserslist": [
                        "chrome >= 87",
                        "edge >= 88",
                        "firefox >= 78",
                        "safari >= 14",
                      ],
                    },
                    "postcssPlugin": "autoprefixer",
                    "prepare": [Function],
                  },
                ],
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "<ROOT>/packages/shared/compiled/less-loader",
            "options": {
              "implementation": "<ROOT>/packages/shared/compiled/less",
              "lessOptions": {
                "javascriptEnabled": true,
              },
              "sourceMap": true,
            },
          },
        ],
      },
      {
        "test": /\\\\\\.toml\\$/,
        "type": "javascript/auto",
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/toml-loader",
          },
        ],
      },
      {
        "test": /\\\\\\.ya\\?ml\\$/,
        "type": "javascript/auto",
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/yaml-loader",
          },
        ],
      },
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/babel-loader/lib/index.js",
            "options": {
              "plugins": [
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "usage",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-react/lib/index.js",
                  {
                    "development": true,
                    "runtime": "classic",
                    "useBuiltIns": true,
                    "useSpread": false,
                  },
                ],
              ],
            },
          },
          {
            "loader": "<ROOT>/node_modules/<PNPM_INNER>/ts-loader/index.js",
            "options": {
              "allowTsInNodeModules": true,
              "compilerOptions": {
                "module": "esnext",
                "target": "esnext",
              },
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "name": "Client",
  "optimization": {
    "minimize": false,
    "runtimeChunk": {
      "name": "bundler-runtime",
    },
    "splitChunks": {
      "cacheGroups": {
        "lib-axios": {
          "name": "lib-axios",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(axios\\|axios-\\.\\+\\)\\[\\\\\\\\/\\]/,
        },
        "lib-lodash": {
          "name": "lib-lodash",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(lodash\\|lodash-es\\)\\[\\\\\\\\/\\]/,
        },
        "lib-polyfill": {
          "name": "lib-polyfill",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(tslib\\|core-js\\|@babel\\\\/runtime\\|@swc\\\\/helpers\\)\\[\\\\\\\\/\\]/,
        },
        "lib-react": {
          "name": "lib-react",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react\\|react-dom\\|scheduler\\)\\[\\\\\\\\/\\]/,
        },
        "lib-router": {
          "name": "lib-router",
          "priority": 0,
          "reuseExistingChunk": true,
          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react-router\\|react-router-dom\\|@remix-run\\\\/router\\|history\\)\\[\\\\\\\\/\\]/,
        },
      },
      "chunks": "all",
      "enforceSizeThreshold": 50000,
    },
  },
  "output": {
    "chunkFilename": "static/js/async/[name].js",
    "filename": "static/js/[name].js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
    "webassemblyModuleFilename": "static/wasm/[hash].module.wasm",
  },
  "performance": {
    "hints": false,
  },
  "plugins": [
    HtmlBasicPlugin {
      "HtmlPlugin": [Function],
      "name": "HtmlBasicPlugin",
      "options": {
        "HtmlPlugin": [Function],
        "info": {},
      },
    },
    DefinePlugin {
      "definitions": {
        "process.env.ASSET_PREFIX": """",
        "process.env.NODE_ENV": ""test"",
      },
    },
    MiniCssExtractPlugin {
      "_sortedModulesCache": WeakMap {},
      "options": {
        "chunkFilename": "static/css/async/[name].css",
        "experimentalUseImportModule": undefined,
        "filename": "static/css/[name].css",
        "ignoreOrder": true,
        "runtime": true,
      },
      "runtimeOptions": {
        "attributes": undefined,
        "insert": undefined,
        "linkType": "text/css",
      },
    },
    HotModuleReplacementPlugin {
      "options": {},
    },
    ReactRefreshPlugin {
      "options": {
        "exclude": [
          /node_modules/,
        ],
        "include": /\\\\\\.\\(\\[cm\\]js\\|\\[jt\\]sx\\?\\|flow\\)\\$/i,
        "overlay": false,
      },
    },
    ProgressPlugin {
      "compileTime": null,
      "dependenciesCount": 10000,
      "handler": [Function],
      "hasCompileErrors": false,
      "id": "Client",
      "modulesCount": 5000,
      "name": "ProgressPlugin",
      "percentBy": null,
      "profile": false,
      "showActiveModules": false,
      "showDependencies": true,
      "showEntries": true,
      "showModules": true,
    },
  ],
  "resolve": {
    "extensions": [
      ".js",
      ".jsx",
      ".mjs",
      ".json",
    ],
  },
  "target": [
    "web",
    "es5",
  ],
}
`;
